File: python/bob/bob.py
import re


def hey(phrase):
    trimmed_phrase = phrase.strip()

    if is_silence(trimmed_phrase):
        return 'Fine. Be that way!'
    elif is_question(trimmed_phrase) and is_yelling(trimmed_phrase):
        return 'Calm down, I know what I\'m doing!'
    elif is_question(trimmed_phrase):
        return 'Sure.'
    elif is_yelling(trimmed_phrase):
        return 'Whoa, chill out!'

    return 'Whatever.'


def is_yelling(phrase):
    return phrase == phrase.upper() and re.search(r'[A-Z]', phrase)


def is_question(phrase):
    return phrase[-1] == '?'


def is_silence(phrase):
    return len(phrase) == 0

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    is_yelling = is_yelling?(input)
    is_question = is_question?(input)

    cond do
      is_question && is_yelling -> "Calm down, I know what I'm doing!"
      is_question -> "Sure."
      is_yelling -> "Whoa, chill out!"
      is_silence?(input) -> "Fine. Be that way!"
      true -> "Whatever."
    end
  end

  defp is_yelling?(normal) do
    upper = String.upcase(normal)
    lower = String.downcase(normal)

    (upper == normal) && (upper != lower)
  end

  defp is_silence?(input) do
    String.trim(input) == ""
  end

  defp is_question?(input) do
    String.ends_with?(input, "?")
  end
end
